plugin/go.vim

```diff
@@ -260,6 +260,16 @@ augroup vim-go
   " in the same window doesn't highlight the most recently matched
   " identifier's positions.
   autocmd BufWinEnter *.go call go#guru#ClearSameIds()
-augroup END
+
+  autocmd BufEnter *.go
+        \  if get(g:, 'go_autodetect_gopath', 0)
+        \|   let b:old_gopath = $GOPATH
+        \|   let $GOPATH = go#path#Detect()
+        \| endif
+  autocmd BufLeave *.go
+        \  if get(g:, 'go_autodetect_gopath', 0)
+        \|   let $GOPATH = b:old_gopath
+        \| endif
+augroup end
```

Comment:
In the rare case when a user changes `g:go_autotdetect_gopath`, `BufLeave` will not behave as expected. Instead of `if get(g: 'go_autodetect_gopath', 0)`, `if exists(b:old_gopath)` will guard restoring `GOPATH` correctly. We should also unset `b:old_gopath`...

Reply:
Good point; I've updated the PR with this issue and the other small issues you found.